File: Thaliak.Service.Poller/Patch/Repository.cs
using System.Text;

namespace Thaliak.Service.Poller.Patch;

public enum Repository
{
    Boot,
    Ffxiv,
    Ex1,
    Ex2,
    Ex3,
    Ex4,
    Ex5
}

public static class RepoExtensions
{
    private static DirectoryInfo GetRepoPath(this Repository repo, DirectoryInfo gamePath)
    {
        switch (repo)
        {
            case Repository.Boot:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "boot"));
            case Repository.Ffxiv:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "game"));
            case Repository.Ex1:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "game", "sqpack", "ex1"));
            case Repository.Ex2:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "game", "sqpack", "ex2"));
            case Repository.Ex3:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "game", "sqpack", "ex3"));
            case Repository.Ex4:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "game", "sqpack", "ex4"));
            case Repository.Ex5:
                return new DirectoryInfo(Path.Combine(gamePath.FullName, "game", "sqpack", "ex5"));
            default:
                throw new ArgumentOutOfRangeException(nameof(repo), repo, null);
        }
    }

    public static FileInfo GetVerFile(this Repository repo, DirectoryInfo gamePath)
    {
        var repoPath = repo.GetRepoPath(gamePath).FullName;
        switch (repo)
        {
            case Repository.Boot:
                return new FileInfo(Path.Combine(repoPath, "ffxivboot.ver"));
            case Repository.Ffxiv:
                return new FileInfo(Path.Combine(repoPath, "ffxivgame.ver"));
            case Repository.Ex1:
                return new FileInfo(Path.Combine(repoPath, "ex1.ver"));
            case Repository.Ex2:
                return new FileInfo(Path.Combine(repoPath, "ex2.ver"));
            case Repository.Ex3:
                return new FileInfo(Path.Combine(repoPath, "ex3.ver"));
            case Repository.Ex4:
                return new FileInfo(Path.Combine(repoPath, "ex4.ver"));
            case Repository.Ex5:
                return new FileInfo(Path.Combine(repoPath, "ex5.ver"));
            default:
                throw new ArgumentOutOfRangeException(nameof(repo), repo, null);
        }
    }

    public static string GetVer(this Repository repo, DirectoryInfo gamePath)
    {
        var verFile = repo.GetVerFile(gamePath);

        if (!verFile.Exists)
            return Constants.BASE_GAME_VERSION;

        var ver =  File.ReadAllText(verFile.FullName);
        return string.IsNullOrWhiteSpace(ver) ? Constants.BASE_GAME_VERSION : ver;
    }

    public static void SetVer(this Repository repo, DirectoryInfo gamePath, string newVer)
    {
        var verFile = GetVerFile(repo, gamePath);

        if (!verFile.Directory.Exists) verFile.Directory.Create();

        using var fileStream = verFile.Open(FileMode.Create, FileAccess.Write, FileShare.None);
        var buffer = Encoding.ASCII.GetBytes(newVer);
        fileStream.Write(buffer, 0, buffer.Length);
        fileStream.Flush();
    }

    public static bool IsBaseVer(this Repository repo, DirectoryInfo gamePath)
    {
        return repo.GetVer(gamePath) == Constants.BASE_GAME_VERSION;
    }
}

File: v2/thaliak-poller/src/patch/repository.rs
use crate::poller::BASE_GAME_VERSION;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GameRepository {
    Boot,
    Ffxiv,
    Ex1,
    Ex2,
    Ex3,
    Ex4,
    Ex5,
}

impl GameRepository {
    fn get_repo_path<P: AsRef<Path>>(&self, game_path: P) -> PathBuf {
        let game_path = game_path.as_ref();
        match self {
            GameRepository::Boot => game_path.join("boot"),
            GameRepository::Ffxiv => game_path.join("game"),
            GameRepository::Ex1 => game_path.join("game").join("sqpack").join("ex1"),
            GameRepository::Ex2 => game_path.join("game").join("sqpack").join("ex2"),
            GameRepository::Ex3 => game_path.join("game").join("sqpack").join("ex3"),
            GameRepository::Ex4 => game_path.join("game").join("sqpack").join("ex4"),
            GameRepository::Ex5 => game_path.join("game").join("sqpack").join("ex5"),
        }
    }

    pub fn get_ver_file<P: AsRef<Path>>(&self, game_path: P) -> PathBuf {
        let repo_path = self.get_repo_path(game_path);
        match self {
            GameRepository::Boot => repo_path.join("ffxivboot.ver"),
            GameRepository::Ffxiv => repo_path.join("ffxivgame.ver"),
            GameRepository::Ex1 => repo_path.join("ex1.ver"),
            GameRepository::Ex2 => repo_path.join("ex2.ver"),
            GameRepository::Ex3 => repo_path.join("ex3.ver"),
            GameRepository::Ex4 => repo_path.join("ex4.ver"),
            GameRepository::Ex5 => repo_path.join("ex5.ver"),
        }
    }

    pub fn get_ver<P: AsRef<Path>>(&self, game_path: P) -> String {
        let ver_file = self.get_ver_file(game_path);

        if !ver_file.exists() {
            return BASE_GAME_VERSION.to_string();
        }

        std::fs::read_to_string(ver_file)
            .unwrap_or_else(|_| BASE_GAME_VERSION.to_string())
            .trim()
            .to_string()
    }

    pub fn is_base_ver<P: AsRef<Path>>(&self, game_path: P) -> bool {
        self.get_ver(game_path) == BASE_GAME_VERSION
    }

    pub fn set_ver<P: AsRef<Path>>(&self, game_path: P, new_ver: &str) -> std::io::Result<()> {
        let ver_file = self.get_ver_file(game_path.as_ref());

        // Create parent directory if it doesn't exist
        if let Some(parent) = ver_file.parent() {
            std::fs::create_dir_all(parent)?;
        }

        std::fs::write(ver_file, new_ver)?;
        Ok(())
    }
}
